### PR TITLE
Implemented shallow device request

### LINKED ISSUE #265: shallow version of DeviceRequest RPC
Our current implementation only provides the deep response.

This'll be a problem for larger models.

The shallow version should:

- deliver the top level of the `Device` i.e. `params`, `commands`, and all the other maps are there but empty.
- stream the contents of each map one at a time
- no need to get more fine-grained than this - i.e. stream top level params separately from their children - until someone proves it's needed

### PR DESCRIPTION
**Pull request by johndanenRV**
closes #265 

Added a coroutine that streams the device as DeviceComponents if shallow is set to true. I wasn't sure where the shallow flag is supposed to come from so, for now, I have hard coded DeviceRequest to always be shallow.

I made sure to add very thorough comments so that it hopefully makes sense to anyone who hasn't worked with coroutines before.

It is interesting to compare this to pull request #170 where I had previously implemented this same feature without coroutines. The logic is clearly much simpler now. 😃

### CODE REVIEW

File: sdks/cpp/connections/gRPC/src/ServiceImpl.cpp [hunk 1]
```diff
@@ -478,20 +478,37 @@ void CatenaServiceImpl::DeviceRequest::proceed(CatenaServiceImpl *service, bool
...
+                } 
+                
+                try {     
+                    catena::DeviceComponent component{};
+                    std::vector<std::string> clientScopes = service_->getScopes(context_);
+                    {
+                        Device::LockGuard lg(dm_);
+                        component = serializer_->getNext();
+                    }
+                    status_ = serializer_->hasMore() ? CallStatus::kWrite : CallStatus::kPostWrite;
```
Comment (mejohnnaylor) [inline sdks/cpp/connections/gRPC/src/ServiceImpl.cpp:503]: shouldn't this line be within the lock guard?

File: sdks/cpp/lite/include/Device.h [hunk 2]
```diff
@@ -146,6 +147,140 @@ class Device {
      */
     void toProto(::catena::LanguageList& list) const;
 
+    /**
+     * @brief DeviceSerializer is a coroutine that serializes the device into a stream of DeviceComponents
+     * This struct manages the state and lifetime of the coroutine. It also provides the interface for resuming the coroutine.
+     */
+    struct DeviceSerializer {
```
Comment (mejohnnaylor) [inline sdks/cpp/lite/include/Device.h:154]: this is  a struct with private attributes - why isn't it a class?

File: sdks/cpp/lite/include/Device.h [hunk 3]
```diff
@@ -146,6 +147,140 @@ class Device {
...
+         * member functions that the compiler will call when the coroutine is created, destroyed,
+         * or resumed.
+         */
+        struct promise_type {
+
+            /**
+             * get_return_object() is called when the coroutine is created (i.e. when getComponentSerializer() 
+             * is called). It creates the handle for the coroutine and returns the DeviceSerializer object.
+             */
+            DeviceSerializer get_return_object() {
```
Comment (mejohnnaylor) [inline sdks/cpp/lite/include/Device.h:183]: should be declared inline or defined in a .cpp file, same comment applied to other methods defined here. recommend inlining them as they're all v short.

**Review by mejohnnaylor — APPROVED** [review]

File: sdks/cpp/connections/gRPC/src/ServiceImpl.cpp (hunk 1 above)
Comment (johndanenRV) [reply sdks/cpp/connections/gRPC/src/ServiceImpl.cpp:503]: No, even if something else changes the device it shouldn't affect how many more components need to be sent.

**Review by mejohnnaylor — APPROVED** [review]

**Pull request merged**